### PR TITLE
Accommodate Java security permissions within Attach API

### PR DESCRIPTION
**Pull request by JasonFengJ9**
#### Accommodate Java security permissions within Attach API ####

To prevent un-intended permission checks on application code, following security sensitive code are protected with `AccessController.doPrivileged`:
1. retrieving system property `com.ibm.tools.attach.timeout`;
2. retrieving system property `com.ibm.tools.attach.command_timeout`;
3. retrieving system property `openj9.tools.attach.diagnostics.debug`;
4. `OpenJ9AttachProvider.listVirtualMachines()`;
5. `OpenJ9VirtualMachine.attachTarget()`;
6. `OpenJ9VirtualMachine.heapHisto()`;

Fixed an API mis-usage `Integer.getInteger(status)`, changed it to `Integer.valueOf(status)`;

Minor code/comment refactoring.

Note: this passes the testcase https://github.com/eclipse/openj9/issues/6119#issuecomment-504538933 with security manager enabled.

Depends: https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/193
Related: https://github.com/eclipse/openj9/pull/6232

Reviewer: @pshipton 
FYI: @pdbain-ibm @DanHeidinga 

Signed-off-by: Jason Feng fengj@ca.ibm.com

### CODE REVIEW

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java [hunk 1]
```diff
@@ -119,6 +121,12 @@
 	/* [PR Jazz 30075] Make syncFileLock an instance variable since it is accessed only by the attachHandler singleton. */
 	FileLock syncFileLock;
 
+	static {
+		shutdownTimeoutMs = AccessController.doPrivileged((PrivilegedAction<Integer>) () -> {
```
Comment (pshipton) [inline jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java]: Do you have a test case showing this change is needed? 
- AttachHander is a bootstrap class used from a separate thread created at startup. I don't think this should be required
- using Lambda's during JVM startup affects startup performance negatively

**Comment by pshipton** [comment]
Seems like a good start, but doesn't fix all the problems previously described. We need a test case that call each of the public APIs.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java (hunk 1 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java]: > Do you have a test case showing this change is needed?

No, the change was purely based on the static code review. Agreed this use scenario doesn't require `doPrivileged()`. Removed the change, please have another look.

Regarding Lambda's usages, a quick search `() ->` shows 600+ hit within `JDK11` extension repo `src` directory while there are other slightly different expressions. Unless there is a critical and significant performance impact, I would like take advantage of Lambda's benefits.

Comment (pshipton) [reply jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java]: See https://github.com/eclipse/openj9/issues/5373
There are no lamba's used in JVM bootstrap. Since you've already removed it, there is no problem.

**Comment by pdbain** [comment]
I believe many of the file access operations are actually performed in the `tools/attach/target` classes.  If these methods are called from `tools/attach/attacher` classes such as `OpenJ9VirtualMachine`, wrapping them in doPriv would remove the need to give `attacher` file access permissions, right?

**Comment by pshipton** [comment]
Yes, but we don't want to add unnecessary doPriv operations. We can use additional test cases to find the additional places where doPriv is required.

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java [hunk 2]
```diff
@@ -81,13 +86,37 @@ public OpenJ9VirtualMachine attachVirtualMachine (
 		}
 
 		OpenJ9VirtualMachine vm = new OpenJ9VirtualMachine(this, descriptor.id());
-		vm.attachTarget();
+		PrivilegedExceptionAction<Object> action = () -> {
+			vm.attachTarget();
```
Comment (pshipton) [inline jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Since attachTarget() is modified to use doPriv, I don't expect this one is needed.

Comment (JasonFengJ9) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: That is true, will remove this doPriv.

Comment (JasonFengJ9) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Removed this `doPrivileged()`.
In addition, modified the testcase https://github.com/eclipse/openj9/issues/6119#issuecomment-504538933 to invoke following public APIs:
```
	vm.getAgentProperties();
	vm.provider();
	vm.startLocalManagementAgent();
	vm.startManagementAgent(props);
	vm.list();
``` 
There was no other `AccessControlException` discovered.

Ideally, existing `AttachAPISanity` test should have a test mode with security manager enabled, will leave that to a separated PR though.

Please review.

Comment (pshipton) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Probably they are fine as well, but there is also the following to check:
detach()
loadAgent()
loadAgentLibrary()
loadAgentPath()
executeDiagnosticCommand()
heapHisto()

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java [hunk 3]
```diff
@@ -88,6 +91,11 @@ public OpenJ9VirtualMachine attachVirtualMachine (
 	@Override
 	public List<VirtualMachineDescriptor> listVirtualMachines() {
 
+		PrivilegedAction<List<VirtualMachineDescriptor>> action = () -> listVirtualMachinesImp();
+		return AccessController.doPrivileged(action);
```
Comment (pshipton) [inline jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: This should re-throw RuntimeException and Error similarly to attachTarget().

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java (hunk 2 above)
Comment (JasonFengJ9) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Good call. Tested following public APIs in modified testcase https://github.com/eclipse/openj9/issues/6119#issuecomment-504538933:
```
	j9vm.getThreadInfo();
	j9vm.heapHisto("-all");
	j9vm.executeDiagnosticCommand("live");
```
Following APIs are tested via `make _TestAttachAPI (org.openj9.test.attachAPI.TestAttachAPI)`:
```
vm.loadAgentLibrary(JVMTITST);
vm.loadAgentLibrary(JVMTITST, "test=aot001");
vm.loadAgentPath(agent);
vm.loadAgentPath(agent, options);
```
`vm.detach()` has already been tested.

Added `doPrivileged()` for `OpenJ9VirtualMachine.heapHisto()`. Please have another look.

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java (hunk 3 above)
Comment (JasonFengJ9) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Fixed.

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9VirtualMachine.java [hunk 4]
```diff
@@ -584,6 +616,23 @@ public String startLocalManagementAgent() throws IOException {
...
+				throw (RuntimeException) cause;
+			} else if (cause instanceof Error) {
+				throw (Error) cause;
+			} else {
+				throw new RuntimeException(cause);
+			}
+		}
+		return ret;
+	}
+	private InputStream heapHistoImpl(Object... opts) {
```
Comment (pshipton) [inline jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9VirtualMachine.java]: Please add a blank line.

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java (hunk 2 above)
Comment (pshipton) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Not sure it matters, but "live" isn't  a valid diagnostic command. Valid command appear to be:
"Thread.print"
"GC.run"
"GC.class_histogram"

Comment (pdbain) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: `live` is a modifier for `heapHisto`.  It causes a `GC.run` before `GC.class_histogram`.

Comment (JasonFengJ9) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9AttachProvider.java]: Those `diagnostic` commands didn't make difference. However just discovered another system property retrieval for `openj9.tools.attach.diagnostics.debug`, which was marked by an excessive custom policy file.

File: jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9VirtualMachine.java (hunk 4 above)
Comment (JasonFengJ9) [reply jcl/src/jdk.attach/share/classes/com/ibm/tools/attach/attacher/OpenJ9VirtualMachine.java]: Fixed. Please review.

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk11 depends ibmruntimes/openj9-openjdk-jdk11#193

**Comment by pshipton** [comment]
Please create the PRs for jdk12, 13, next.
Also this OpenJ9 PR against the 0.15 branch.
